### PR TITLE
Immutable HttpEntity classes

### PR DESCRIPTION
**Pull request by ok2c**

### CODE REVIEW

File: httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/InputStreamEntity.java [hunk 1]
```diff
@@ -31,87 +31,52 @@
...
 import org.apache.hc.core5.util.Args;
 
 /**
- * A streamed, non-repeatable entity that obtains its content from
- * an {@link InputStream}.
+ * A streamed, non-repeatable entity that obtains its content from an {@link InputStream}.
  *
  * @since 4.0
  */
+@Contract(threading = ThreadingBehavior.IMMUTABLE_CONDITIONAL)
```
Comment (rschmitt) [inline httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/InputStreamEntity.java]: I don't really see how an `InputStream` wrapper could be regarded as immutable, unless `IMMUTABLE_CONDITIONAL` specifically means "immutable" in the Java Memory Model sense (all final fields, not made visible to other threads during construction), which has significantly weaker guarantees than true immutability.

File: httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/SerializableEntity.java [hunk 2]
```diff
@@ -59,29 +62,35 @@
...
-    public SerializableEntity(final Serializable ser, final boolean bufferize) throws IOException {
-        super();
+    public SerializableEntity(
+            final Serializable ser, final boolean bufferize, final ContentType contentType,
+            final String contentEncoding) throws IOException {
+        super(contentType, contentEncoding);
         Args.notNull(ser, "Source object");
         if (bufferize) {
             createBytes(ser);
+            this.objRef = null;
```
Comment (rschmitt) [inline httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/SerializableEntity.java]: I don't understand what's happening here. Why are you nulling this field out after `createBytes` has assigned it?

File: httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/SerializableEntity.java [hunk 3]
```diff
@@ -91,33 +100,33 @@ private void createBytes(final Serializable ser) throws IOException {
     }
 
     @Override
-    public InputStream getContent() throws IOException, IllegalStateException {
+    public final InputStream getContent() throws IOException, IllegalStateException {
         if (this.objSer == null) {
             createBytes(this.objRef);
```
Comment (rschmitt) [inline httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/SerializableEntity.java]: Based on this (apparent) lazy initialization path, this class is not currently immutable in any sense.

File: httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/InputStreamEntity.java (hunk 1 above)
Comment (ok2c) [reply httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/InputStreamEntity.java]: @rschmitt This is why there is `IMMUTABLE` and `IMMUTABLE_CONDITIONAL` with much weaker guarantees.

File: httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/SerializableEntity.java (hunk 2 above)
Comment (ok2c) [reply httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/SerializableEntity.java]: @rschmitt Neither do it. I did not author this class.

File: httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/SerializableEntity.java (hunk 3 above)
Comment (ok2c) [reply httpcore5/src/main/java/org/apache/hc/core5/http/io/entity/SerializableEntity.java]: @rschmitt I removed `IMMUTABLE_CONDITIONAL` annotation from `ByteBufferEntity` and `InputStreamrEntity` and removed buffering code from `SerializableEntity`.

**Pull request merged**